light-client/src/builder/light_client.rs

```diff
@@ -0,0 +1,218 @@
+pub struct NoTrustedState;
+
+/// A trusted state has been set and validated
+pub struct HasTrustedState;
+
+/// Builder for a light client [`Instance`]
+#[must_use]
+pub struct LightClientBuilder<State> {
+    peer_id: PeerId,
+    options: Options,
+    io: Box<dyn Io>,
+    clock: Box<dyn Clock>,
+    hasher: Box<dyn Hasher>,
+    verifier: Box<dyn Verifier>,
+    scheduler: Box<dyn Scheduler>,
+    predicates: Box<dyn VerificationPredicates>,
+    light_store: Box<dyn LightStore>,
+
+    #[allow(dead_code)]
+    state: State,
+}
+
+impl<Current> LightClientBuilder<Current> {
+    /// Private method to move from one state to another
+    fn with_state<Next>(self, state: Next) -> LightClientBuilder<Next> {
+        LightClientBuilder {
+            peer_id: self.peer_id,
+            options: self.options,
+            io: self.io,
+            clock: self.clock,
+            hasher: self.hasher,
+            verifier: self.verifier,
+            scheduler: self.scheduler,
+            predicates: self.predicates,
+            light_store: self.light_store,
+            state,
+        }
+    }
+}
+
+impl LightClientBuilder<NoTrustedState> {
+    /// Initialize a builder for a production (non-mock) light client.
+    #[cfg(feature = "rpc-client")]
+    pub fn prod(
+        peer_id: PeerId,
+        rpc_client: rpc::HttpClient,
+        light_store: Box<dyn LightStore>,
+        options: Options,
+        timeout: Option<Duration>,
+    ) -> Self {
+        Self::custom(
+            peer_id,
+            options,
+            light_store,
+            Box::new(ProdIo::new(peer_id, rpc_client, timeout)),
+            Box::new(ProdHasher),
+            Box::new(SystemClock),
+            Box::new(ProdVerifier::default()),
+            Box::new(scheduler::basic_bisecting_schedule),
+            Box::new(ProdPredicates),
+        )
+    }
+
+    /// Initialize a builder for a custom light client, by providing all dependencies upfront.
+    #[allow(clippy::too_many_arguments)]
+    pub fn custom(
+        peer_id: PeerId,
+        options: Options,
+        light_store: Box<dyn LightStore>,
+        io: Box<dyn Io>,
+        hasher: Box<dyn Hasher>,
+        clock: Box<dyn Clock>,
+        verifier: Box<dyn Verifier>,
+        scheduler: Box<dyn Scheduler>,
+        predicates: Box<dyn VerificationPredicates>,
+    ) -> Self {
+        Self {
+            peer_id,
+            hasher,
+            io,
+            verifier,
+            light_store,
+            clock,
+            scheduler,
+            options,
+            predicates,
+            state: NoTrustedState,
+        }
+    }
+
+    /// Set the given light block as the initial trusted state.
+    pub fn trust_light_block(
+        mut self,
+        trusted_state: LightBlock,
+    ) -> Result<LightClientBuilder<HasTrustedState>, Error> {
+        self.validate(&trusted_state)?;
+
+        // TODO(liamsi, romac): it is unclear if this should be Trusted or only Verified
```

Comment:
it may be a good idea to cross-verify the first header to check if all witnesses have the same view.

but in the end, you have to trust the state I think.

Reply:
> It may be a good idea to cross-verify the first header to check if all witnesses have the same view.

Sounds like a good idea, indeed!

@ebuchman What do you think? Should we address this in a follow-up PR?

Reply:
Yup, if anyone disagrees should probably throw an error and refuse to start

Reply:
yeah I agree. We currently just check that the chain ID's are the same but it would make more sense to cross-verify the first header. This way we can also check for responsiveness i.e. if a witness doesn't respond we drop them, and for correctness i.e. if one of the headers are different we halt and return the error

Reply:
I have collected the outcome of the discussions regarding initialization into
https://github.com/tendermint/spec/pull/159

We have to trust the given initial header. Cross-check doesn't do anything useful. It only opens the possibility that one fault peer can prevent the light node to start.

Reply:
Might as well catch that error at startup and get the user to provide a different witness rather than catch it later and just drop the witness without replacement, no ?

Reply:
> Might as well catch that error at startup and get the user to provide a different witness rather than catch it later and just drop the witness without replacement, no ?

We can do that. Still, the user might need to do that several times, if the user is unlucky and always chooses another faulty peer. And these faulty peers don't even need to be clever but just return garbage with the right height. Still, if it makes sense in the practical setting we can do that. It is definitely safe.